week5/Calc.hs

```diff
@@ -0,0 +1,95 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+module Calc where
+
+import ExprT
+import Parser
+import StackVM
+import Data.Maybe
+import qualified Data.Map as M
+
+-- Exercise 1
+eval :: ExprT -> Integer
+eval (ExprT.Lit x) = x
+eval (ExprT.Add x y) = eval x + eval y
+eval (ExprT.Mul x y) = eval x * eval y
+
+-- Exercise 2
+evalStr :: String -> Maybe Integer
+evalStr x = case parseExp ExprT.Lit ExprT.Add ExprT.Mul x of
+            Just y -> Just (eval y)
+            _ -> Nothing
+
+-- Exercise 3
+class Expr a where
+  lit :: Integer -> a
+  mul :: a -> a -> a
+  add :: a -> a -> a
+
+instance Expr ExprT where
+  lit = ExprT.Lit
+  mul = ExprT.Mul
+  add = ExprT.Add
+
+-- Exercise 4
+instance Expr Integer where
+  lit = id
+  mul = (*)
+  add = (+)
+
+instance Expr Bool where
+  lit = (>0)
+  mul = (&&)
+  add = (||)
+
+newtype MinMax = MinMax Integer deriving (Eq, Show)
+newtype Mod7   = Mod7 Integer deriving (Eq, Show)
+
+instance Expr MinMax where
+  lit = MinMax
+  mul (MinMax x) (MinMax y) = MinMax (if x < y then x else y)
+  add (MinMax x) (MinMax y) = MinMax (if x > y then x else y)
+
+instance Expr Mod7 where
+  lit = Mod7
+  mul (Mod7 x) (Mod7 y) = Mod7(mod (x*y) 7)
+  add (Mod7 x) (Mod7 y) = Mod7(mod (x+y) 7)
+
+-- Exercise 5
+instance Expr StackVM.Program where
+  lit x = [StackVM.PushI x]
+  add x y = x ++ y ++ [StackVM.Add]
+  mul x y = x ++ y ++ [StackVM.Mul]
```

Comment:
We also use 4 space-indentation in our style guide 🙂

Reply:
Unfortunately, it's only me and you who follow our style guide. This brings the question if the company should rewrite the style-guide or to agree that there is no any.

Reply:
I had this exact same thing happened with previous iOS team, and we end up with 2 spaces.
But if majority prefer 4 spaces, I'll abide. 👍

Reply:
Funny story is that we have a style guide exactly for the reasons that we don't need to end up arguing with the team. But as I said if the majority is not using it so I guess you can do whatever you want too.